### ISSUE TITLE
Incorrect poll() handling in getreq() for keep alive connections.

### ISSUE THREAD

**Issue by sfyang** [bel2125/civetweb#82]
Recently I used civetweb as an embedded HTTP server in a project that requires the support of persistent (keep alive) connections, and I found that there seems to be some glitches in the polling code in the getreq() function.

In `process_new_connection()`, function `getreq()` is called in the beginning of the `do ... while` loop with timeout set to `TIMEOUT_INFINITE`, which is defined as an enum constant with value `-1` in `civetweb.h`. However, in the beginning of `getreq()`, there is a code segment dealing with polling from the client socket that has unreachable `poll()` call in it:

```C
if (timeout >= 0) {
    pfd.fd = conn->client.sock;
    switch (poll(&pfd, 1, timeout)) {
    case 0:
        snprintf(ebuf, ebuf_len, "%s", "Timed out");
        *err = 408;
        return 0;
    case -1:
        snprintf(ebuf, ebuf_len, "%s", "Interrupted");
        *err = 500;
        return 0;
    }
}
```

Since `getreq()` is called with timeout set to `TIMEOUT_INFINITE` that is equal to -1, the control will never reach the `poll()` call. Besides, even if the control could reach the `poll()` call, since there is no any event registered to the `pfd.events` field, the `poll()` call will block the execution until either timed out or being interrupted.

The improper `poll()` handling would probably not be a big problem for non-persistent connections since the client socket would most likely be ready to be read when `getreq()` is first called. However, it could cause problems for persistent connections when `getreq()` is called the next time in the `do ... while` loop if the client idles after the first request.

When `getreq()` is called after the loop starts again, since `poll()` call is skipped, the following `read_request()` then calls the `pull()` function, which in turn tries to read from underlying I/O channel and block the execution until something is read, the underlying I/O call (either `recv()` or `SSL_read()`) is interrupted, or the I/O call is timed out. If the client idles and the low level I/O call is timed out, the `read_request()` call returns error, and a 400 message is sent to the client before the connection is closed.

To demonstrate this problem, compile and execute the "civetweb" server in one console as follow:
```bash
$ echo 'Hello World' > hello.txt
$ ./civetweb -enable_keep_alive yes -document_root .
Civetweb V1.7 started on port(s) 8080 with web root [/tmp/civetweb-test]
```
and connect to localhost:8080 with telnet in another console window, send "GET /hello.txt" request with "Connection: Keep-Alive" header and then idle for more than 30 seconds:
```bash
$ telnet localhost 8080
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
GET /hello.txt HTTP/1.1
Connection: Keep-Alive

HTTP/1.1 200 OK
Date: Mon, 23 Mar 2015 15:43:20 GMT
Last-Modified: Mon, 23 Mar 2015 15:39:04 GMT
Etag: "55103398.12"
Content-Type: text/plain
Content-Length: 12
Connection: keep-alive
Accept-Ranges: bytes

Hello World
HTTP/1.1 400 Bad Request
Content-Type: text/plain
Content-Length: 52
Date: Mon, 23 Mar 2015 15:43:50 GMT
Connection: keep-alive

Error 400: Bad Request
Client sent malformed requestConnection closed by foreign host.
```

As shown above, an unexpected 400 response was sent by the server, and then the connection was terminated.

For now, I circumvent this problem by removing the `if (timeout >= 0)` check, and register `POLLIN` to the `pfd.events` field. In this way, the `poll()` call will poll the client socket with infinite timeout. I wonder if there is better solution to this, such as, instead of calling `getreq()` with `TIMEOUT_INFINITE` unconditionally, calling it with the timeout value of the "request_timeout_ms" config option when "enable_keep_alive" is set to "yes", and close connection without sending 400 response if the `poll()' call times out.

Please advise. Thanks!

**Comment by bel2125** [comment]
Currently, there is also another timeout issue (#77, https://github.com/bel2125/civetweb/issues/77), so I probably have to take some time to check the entire timeout logic.

For this point: right, TIMEOUT_INFINITE should be directly passed to poll, since it will also consider negative values as infinite (http://linux.die.net/man/2/poll). It looks like pfd.events is not actually 0 but uninitialized. I wonder why this did not cause any problems earlier.
I will have a look.

**Comment by bel2125** [comment]
The poll call does not make any sense at this point - neither with nor without timeout.
A timeout needs to be implemented differently.

**Comment by sfyang** [comment]
Hello,

I saw that now the timeout handling is all done in the read_request(). It looks like the while loop in read_request() handles timeout **within** a single request by checking the time difference between consecutive pull() calls. However, I think the timeout for a keep_alive connection **between** two requests should be handled separately.

In current implementation, the testing case I posted can still be triggered, and an unexpected 400 response is still sent if the client idles more than REQUEST_TIMEOUT seconds between requests since pull() returns -1 in such a case. It probably would be better to have another timeout config options for specifying the connection timeout for the keep_alive connections, and the server would just close the connection without sending 400 status to the client if client idles longer than the timeout value specified by the between-request timeout config.

Thanks!

PS. By the way, there is a misspelling for the variable name "request_**timout**" in read_request().
